Paper: Monaco: A Monte Carlo Library for Performing Uncertainty and Sensitivity Analyses

scottshambaugh opened the pull request:
Please note that I have some TODOs as indicated (mostly fleshing out references), but the text and images are my completed draft. Hope that's okay, I'll try to get all the TODOs wrapped up in the next week.

https://procbuild.scipy.org/
https://procbuild.scipy.org/download/scottshambaugh-shambaugh

*Editor*: <!--editor-->@mepa<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@csadorf, @shubhamsharma1609<!--end-reviewers-list-->

scottshambaugh committed:
Init paper

scottshambaugh committed:
paper WIP

scottshambaugh committed:
more WIP!

scottshambaugh committed:
Draft paper, still some small TODOs

scottshambaugh committed:
Remove extra files

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

scottshambaugh committed:
Tweaks

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"scottshambaugh-shambaugh"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmprb8igg99'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: scott_shambaugh\n*** Warning: /tmp/tmprb8igg99/publisher/../output/scott_shambaugh/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/scottshambaugh-shambaugh.pdf", "build_status"=>"success", "build_timestamp"=>"28/05 05:42"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @mepa as editor

scoobies commented:
Assigned! @mepa is now the editor

deniederhut commented:
@scoobies add @csadorf to reviewers

scoobies commented:
@csadorf added to the reviewers list!

scottshambaugh committed:
Set up separate .bib file

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

deniederhut commented:
@scoobies add @shubhamsharma1609 to reviewers

scoobies commented:
@shubhamsharma1609 added to the reviewers list!

mepa commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper scottshambaugh-shambaugh scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

mepa commented:
> Please note that I have some TODOs as indicated (mostly fleshing out references), but the text and images are my completed draft. Hope that's okay, I'll try to get all the TODOs wrapped up in the next week.

Hi @scottshambaugh! Although I see that the paper is ready for review of text and images, it would be great if you can also wrap up the TODOs in the coming days so that references, etc. are available to reviewers as they work through their initial reviews. Thanks!

scottshambaugh commented:
Apologies @mepa, will wrap it up on Sunday.

scottshambaugh committed:
Resolve TODOs

scottshambaugh committed:
Final draft

scottshambaugh commented:
@mepa TODOs are resolved and draft is in it's final state ready for review.

scottshambaugh commented:
@scoobies build paper

scottshambaugh commented:
@scoobies check references

scottshambaugh committed:
layout

scottshambaugh committed:
reference

mepa commented:
Hi @csadorf and @shubhamsharma1609! A friendly reminder that the deadline for first reviews is this Thursday, June 16th.

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+Introduction
+============
+
+Computational models form the backbone of decision-making processes in science, engineering, and public policy. However, our increased reliance on these models stands in contrast to the difficulty in understanding them as we add increasing complexity to try and capture ever more of the fine details of real-world interactions. Practitioners will often take the results of their large, complex model ...
```
While potentially backed up later, this paragraphs contains a lot of unsubstantiated claims about the field and common practices. Would it be possible to provide references right away?

scottshambaugh replied:
Added earlier reference.

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+
+   The monaco project logo. :label:`monacologo`
+
+Why Monte Carlos?
```
This sounds odd, what are "Monte Carlo_s_"?
```suggestion
Motivation for Monte Carlo approach
```

scottshambaugh replied:
Changed to suggestion

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+
+A complete exploration of a model's input space is necessary to fully characterize its behavior, and responsible use of computational models requires understanding the total range of predictions they imply. Policy papers have identified UA and SA as a critical good modeling practice :cite:`azzini2020uncertainty` :cite:`us2009guidance`. The short answer to the question of when to conduct UA and SA ...
+
+There are some important considerations to keep in mind, however. With computational power making running large numbers of cases ever easier, a Monte Carlo analysis can result in highly statistically signficant conclusions. However these results are *conditional on the correctness of the underlying model and input distributions*. If the underlying model has not been throroughly validated, then an ...
```
This and the next paragraph appear a bit "self-evident" and do not provide much substance to the paper in my opinion.

scottshambaugh replied:
Cut this down significantly.

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+
+This validation can be difficult - outputs are a function of the combined inputs and model, and cannot be used to validate either the inputs or the model on their own. Generally validation requires significant domain expertise to ensure a mechanistic model has a solid theoretical foundation, and to diagnose errors in its implementation. 
+
+This is not to say that UA and SA should not be conducted early in the model development process - obtaining the range of plausible output uncertainties is a critical step in input and model validation. Test data cannot be well compared against a single point estimate of a model's output, and it is necessary to have the full distribution of output values to compare test data against. Once a Monte ...
```
This and the preceeding paragraphs are a bit too much "waffle". The paper would benefit from staying more focused on the topic at hand, the particular software that is presented, and stay clear  of making an attempt at providing a holistic motivation for the underlying methodology.

scottshambaugh replied:
Cut this down significantly

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+
+One inherent pitfall of Monte Carlo approaches is that rare events may be undersampled. For example, NASA uses Monte Carlo simulations extensively during launch vehicle design to predict the rocket trajectory and performance. :cite:`hanson2010applying` However, they must prove robustness to anomalous or stressing scenarios which may occur only one or two times in a run of thousands of cases, whic ...
+
+Note that *monaco*'s computational and storage overhead in creating easily-iterrogatable objects for each variable, value, and case makes it an inefficient choice for computationally simple applications with high :math:`n`, such as Monte Carlo integration. It is best suited for models with moderate to high computational cost. 
```
So, for what range of _n_ is it suitable? What would be considered moderate to high computational costs? What is the computational complexity? Are there any benchmarks?

scottshambaugh replied:
Added some numbers on storage and computational costs, this is going to be very model dependent however

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+Why Use Basic Monte Carlo over Bayesian Methods?
+------------------------------------------------
+
+*monaco* purposefully eschews the greater computational efficiency and built-in input parameter estimation of Bayesian methods such as Markov Chain Monte Carlo in favor of basic sampling. For its target audience of scientists, engineers, and policy analysts with mechanistic models, this is useful for many reasons:
```
I think it is borderline absurd to claim that basic techniques such as MCMC would be too complex for scientists, engineers, and policy analysists to understand.

Already considering the next paragraph, it is absolutely possible for software with well-designed API to be suitable to both novice and advanced users. Using layered APIs and appropriate defaults, complexity can be effectively hidden, while still empowering the expert user to embrace them when needed.

scottshambaugh replied:
Cut this down significantly, will address your point in separate comment

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+
+*monaco* purposefully eschews the greater computational efficiency and built-in input parameter estimation of Bayesian methods such as Markov Chain Monte Carlo in favor of basic sampling. For its target audience of scientists, engineers, and policy analysts with mechanistic models, this is useful for many reasons:
+
+* `monaco` favors conceptual simplicity for greater accessibility. The goal is to be usable by someone at the level of knowing what a Uniform and Normal distribution are, and not require an in-depth data science background in inference. For one example of a barrier to entry that would not be covered in an undergraduate education, Bayesian methods generate inherently correlated sample points, whic ...
```
I would have expected that a tool that claims to be a "Monte Carlo _Framework_" would at least provide the option for something other than basic random sampling.

scottshambaugh replied:
Didn't realize the extent to which framework was a term of art! Changed to "library" everywhere

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+Overall Structure
+-----------------
+
+Broadly, each input factor and model output is a *variable* that can be thought of as lists (rows) containing the full range of randomized *values*. *Cases* are slices (columns) that take the *i*'th input and output value for each variable, and represent a single run of the model. Each case is run on its own in parallel, and the output values are collected into output variables. Fig. :ref:`figarc ...
```
What parallelization technique is used?

It is clarified later that the software uses dask for parallel processing, might be worthwhile to at least mention here.

scottshambaugh replied:
I just took it out of this section

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+Further Information
+===================
+
+*monaco* is available on PyPI, has API documentation at https://monaco.rtfd.io/, and is hosted on github at https://github.com/scottshambaugh/monaco/. Please see the "examples" directory in the github source for many more Monte Carlo implementation examples across a range of domains such as physics simulation, election prediction, financial modeling, pandemic spread, and integration.
```
I'd suggest to mention that is available on PyPI under the same name (`monaco`). While this may seem obvious, this just removes any doubt that this is the right package.

scottshambaugh replied:
Good idea, done

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+Sampling Methods
+----------------
+
+Random sampling of the percentiles for each variable can be done using scipy's pseudo-random number generator (PRNG), or with any of the low-discrepancy methods in `scip.stats.qmc` Quasi-Monte Carlo module. In general, the `'sobol_random'` method that generates Sobol sequences with Owen scrambling :cite:`sobol1967distribution` :cite:`owen2020dropping` is recommended in nearly all cases as a well- ...
```
Using Sobol sequences, wouldn't this classify the algorithm as Quasi-Monte Carlo?

Further, looking at the uniform distribution and the frequency spectrum plot, there are some very clear artifacts that I would expect to lead to biased sampling with respect to specific intervals. Maybe you can comment on the effect/non-effect of those correlations on the UA/SA results?

scottshambaugh replied:
Yes, made the QMC distinction more explicit. Added a note on the spectra and some more justification for the default being scrambled Sobol sampling

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,337 @@
+*monaco* purposefully eschews the greater computational efficiency and built-in input parameter estimation of Bayesian methods such as Markov Chain Monte Carlo in favor of basic sampling. For its target audience of scientists, engineers, and policy analysts with mechanistic models, this is useful for many reasons:
+
+* `monaco` favors conceptual simplicity for greater accessibility. The goal is to be usable by someone at the level of knowing what a Uniform and Normal distribution are, and not require an in-depth data science background in inference. For one example of a barrier to entry that would not be covered in an undergraduate education, Bayesian methods generate inherently correlated sample points, whic ...
+* Many modeling domains are not data-rich, especially in contrast to the dimensionality :math:`k` of the model. This precludes using standard Bayesian approaches to obtain results with any useful level of confidence. Consider NASA's use of Monte Carlo in predicting a rocket's flight :cite:`hanson2010applying` - they might have high levels of confidence that the hundreds of parameters describing t ...
```
I feel like these broad claims about various modeling domains are overly reductive.

shubham-scisar replied:
@scottshambaugh  I have gone through the paper and the paper seems to be a thorough effort at developing a comprehensive library to support monte carlo simulations. The paper elaborates very well the working of the library and focusses on the key objects via examples. At a macro level, I can say that the paper fulfils all the key aspects that it should cover. At micro level analysis with separate sections under consideration,I would strongly suggest, to add a few case studies (one or two in no.) application areas where monte carlo simulations are used. For example, it is mentioned that NASA utilises monte carlo simulations for rocket trajectory predictions,it would be better if this or another case study is implemented utilising this library. This will enhance the purpose of library and motivate the readers to experiment with it. Kindly incorporate.

scottshambaugh replied:
csadorf, reworked this expensively and removed claims about modeling domains.
@shubhamsharma1609, I've added two case studies to the end of the paper

csadorf reviewed (changes requested):
This paper introduces the software `monaco` and provides motivation for both the foundation and application of the methods that it implements (UA and SA via Monte-Carlo sampling). It further provides a broader review of the state of the field and makes an appeal to the reader to employ the discussed the techniques in almost all instances where they are technically applicable.

The software itself appears to be professionally developed and maintained. It has a comprehensive README and documentation, an open-source license, is distributed via standard mechanisms (PyPI), takes advantage of state of the art libraries within the scientific Python eco system (numpy, scipy, dask, to name just a few), provides many examples, and appears to be well-tested. The code quality appears to be high, the scholarly effort with about 2800 lines of code is slightly questionable.

The language of the paper is clear, consistent, and mostly precise. The figures are of very high quality and support the narrative in a constructive manner. The choice of examples is appropriate and helps to increase the understanding of the scope and usage patterns of the presented software.

Despite this, I am not convinced that the paper can be published in the currently presented form.
The paper lacks focus and would greatly benefit from a major revision that sharpens its focus on the presented tool rather than making an attempt at providing a holistic review of the foundation and motivation for UA and SA as well as MC sampling in general. A large portion of the introduction and first chapter is spent on discussing how UA/SA is and is not used in various fields, where it would be applicable and where not, and how the use of MC techniques in general is justified. Especially with the conference audience in mind, I would think that at most 3-4 paragraphs that give a very brief summary and motivation and otherwise direct the reader to the relevant literature would be appropriate and more effective.

A significant portion is dedicated to justifying why monaco does not implement any Bayesian methods, but rather sticks to basic sampling. One provided reason is that importance sampling would be too complex for the target audience to understand. In light that the stated target audience is made up of "scientists, enginers, and policy analysists" makes this claim borderline absurd and thus weakens the paper significantly. Irrespective of the need for a safe "default", the choice of the appropriate sampling method should in the end lie with the user/practitioner. Whether the method is used appropriately or not should be up to their judgement – claiming to protect the user from making a mistake by using the wrong method by simply not implementing it is imposing at best. Instead, the author could simply state that uniform sampling is a safer choice and was therefor implemented (first) and that the implementation of Bayesian methods is either out-of-scope or might be left to future development.

The lack of choice with respect to sampling methods may or may not be justified, however I believe that the main weakness of the tool lies within its execution model. I think in the majority of cases, users would prefer to perform the MC sampling of their respective model using existing tools and methods, and then feed the outputs to monaco for analysis rather than being forced to wrap their models in Python functions with all the inherent drawbacks. If an "online" analysis is strictly necessary, a better model in my opinion, would be for the user to request the random variables from monaco, apply them in a way that is suitable, and then feed the outputs back.

I would request for the author to comment on some of the weaknesses that I have pointed out, and recommend to revise the paper to increase its focus on the tool at hand and to either remove or better substantiate some of the claims I have critized in my review.

deniederhut commented:
@scoobies mark needs more review

scottshambaugh commented:
Thank you for the review @csadorf! A lot to chew on, and I really appreciate the thought that went into it. I believe I should be able to address all the issues you brought up, but will wait for the other reviewer to complete their look before responding in detail.

csadorf commented:
> Thank you for the review @csadorf! A lot to chew on, and I really appreciate the thought that went into it. I believe I should be able to address all the issues you brought up, but will wait for the other reviewer to complete their look before responding in detail.

Great, sounds good! Looking forward to it! 😊👍

scottshambaugh commented:
Thank you for the review @shubhamsharma1609! At the end of the paper I point to the project github repo with [worked examples](https://github.com/scottshambaugh/monaco/tree/main/examples) across several domains. Does that satisfy your request? I think that these are generally a little too involved to include all the code for in the paper.
Perhaps four "teaser" plots in a single 2x2 figure from those examples would be helpful to include at the end? Though I don't expect I would be able to go into details on how each were generated. Or perhaps focusing in on the baseball example at a higher level.

deniederhut commented:
@scoobies mark pending comment

scottshambaugh committed:
Reviewer updates

scottshambaugh committed:
Tweaks

scottshambaugh commented:
@csadorf I've reworked the paper to address your concerns. It was incredibly helpful getting outside eyes on it and I think the changes you proposed really do help with tightening it up. To address your points individually:

> The paper lacks focus and would greatly benefit from a major revision that sharpens its focus on the presented tool rather than making an attempt at providing a holistic review of the foundation and motivation for UA and SA as well as MC sampling in general. [...]

I took out most of the foundation and motivation for the methods and pointed towards some more references. Added in some more technical feature subsections as well. Wanted to keep most of the "mathematical grounding" section in since it sets up some variables that I reference throughout, but shortened that a bit as well.

> A significant portion is dedicated to justifying why monaco does not implement any Bayesian methods, but rather sticks to basic sampling. One provided reason is that importance sampling would be too complex for the target audience to understand. In light that the stated target audience is made up of "scientists, enginers, and policy analysists" makes this claim borderline absurd and thus weakens the paper significantly. [...]
> [...] it is absolutely possible for software with well-designed API to be suitable to both novice and advanced users. [...]

My intent was not to say that other sampling methods are too complex for the target audience to understand, but that such methods raise the barrier to entry. More complex tools exist, and the lack of a tool that really focuses in on basic Monte Carlo sampling is still I think the most valuable gap this tool fills. In my professional experience being able to teach new engineers quickly and communicate methodology to a less technical audience has been one of the biggest barriers to adoption of techniques like this. 

However, points taken on hiding complexity as well as the broader point of not imposing a methodology on the reader based on one's own value judgement. The latter doesn't really belong in a paper. This section was removed completely, and replaced with a paragraph describing the scope of the library and some pointing to other tools the reader might want to use with more complexity.

> The lack of choice with respect to sampling methods may or may not be justified, however I believe that the main weakness of the tool lies within its execution model. […]

This was a use case I hadn't at all considered, but makes a lot of sense and thank you for the suggestion! Importing and exporting variables to file was a pretty easy feature to add, and the library should be able to accommodate these workflows now. I added a subsection describing this.

Please take another look and let me know if these changes satisfy your concerns with the paper, I really appreciate the thorough review and am happy to make further revisions if needed!

scottshambaugh committed:
Plot tweak

csadorf commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper scottshambaugh-shambaugh scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

scottshambaugh commented:
Original first draft, for reference if it's easier than looking at diffs: 
[paper.pdf](https://github.com/scipy-conference/scipy_proceedings/files/8985517/paper.pdf)

deniederhut commented:
Hi @csadorf and @shubhamsharma1609! Do you feel this paper is now ready for inclusion in the proceedings?

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,400 @@
+
+UA and SA of any model follows a common workflow. Probability distributions for the model inputs are defined, and randomly sampled values for a large number of cases are fed to the model. The outputs from each case are collected and the full set of inputs and outputs can be analyzed. Typically UA is performed by generating histograms, scatter plots, and summary statistics for the output variables ...
+
+Note again that *monaco* does not currently have tools for model or input parameter validation, and closing that part of the workflow loop is left up to the user. 
```
```suggestion
Model or input parameter validation does not fall into *monaco*'s scope; closing that part of the workflow loop is left up to the user.
```

scottshambaugh replied:
Reworded

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,400 @@
+2) Add input variables to the sim with specified probability distributions.
+3) Run the simulation. This executes the following:    
+
+ a) Random percentiles :math:`p_i \in U[0,1]` are drawn `ndraws` times for each of the input variables.
```
This is rendered oddly.

scottshambaugh replied:
I have not been able to figure out how to get it to indent normally. @mepa?

scottshambaugh replied:
Opened an issue here: https://github.com/scipy-conference/scipy_proceedings/issues/741

deniederhut replied:
The rst indentation is three spaces, not one:

<img width="421" alt="Screen Shot 2022-06-28 at 8 01 37 PM" src="https://user-images.githubusercontent.com/6868396/176328936-96a22b75-d292-4b6f-95ea-f0e2cfca8ed2.png">

More examples here: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html

scottshambaugh replied:
Thank you!

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,400 @@
+Incorporating into Existing Workflows 
+-------------------------------------
+
+If the user want to use existing workflows for generating, running, post-processing, or examining results, any combination of *monaco*'s major steps can be replaced by external tooling by saving and loading input and output variables to file. For example, *monaco* can be used only for its parallel processing backend by importing existing randomly drawn input variables, running the simulation, and ...
```
wants*

scottshambaugh replied:
Fixed

csadorf commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,400 @@
+
+A rigorous way of choosing the number of cases is to first choose a statistical interval (e.g. a confidence interval for a percentile, or a tolerance interval to contain a percent of the population), and then use order statistics to calculate the minimum :math:`n` required to obtain that result at a desired confidence level. This approach is independent of :math:`k`, making UA of high-dimensional ...
+
+A more qualitative UA method would simply to choose a reasonably high :math:`n` (say, :math:`n=2^{10}`), manually examine the results to ensure high-interest areas are not being undersampled, and rely on bootstrapping of the desired variable statistics to obtain the required confidence levels. 
```
would simply choose*

scottshambaugh replied:
Reworded

csadorf reviewed (approved):
Thank you very much for addressing my comments! The paper is ready to be included in the proceedings in my opinion.

I found a few typo and minor grammar issues while reading. I suggest that you have another very close look at the proof and fix those.

scottshambaugh committed:
Tweaks

scottshambaugh committed:
Spacing

mepa commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1002/9780470316771.ch5 is OK
- 10.2514/6.2010-8433 is OK
- 10.25080/majora-7b98e3ed-013 is OK
- 10.1016/j.envsoft.2019.01.012 is OK
- 10.1029/2020gl089829 is OK
- 10.14293/s2199-1006.1.sor-life.a7056644.v1.rysreg is OK
- 10.7717/peerj-cs.55 is OK
- 10.1016/j.envsoft.2020.104954 is OK
- 10.1111/cgf.13366 is OK
- 10.48550/arXiv.2008.08051 is OK
- 10.1017/CBO9780511761188 is OK
- 10.1137/070709359 is OK
- 10.1002/rsa.3240030107 is OK
- 10.1002/2015wr017558 is OK
- 10.1002/9780470725184 is OK
- int.j.uncertaintyquantification.2021038133 is OK
- 10.2760/922129 is OK
- 10.1016/j.envsoft.2020.104954 is OK
- 10.1016/j.scitotenv.2016.02.133 is OK
- 10.1016/j.envsoft.2019.104588 is OK
- 10.1017/S0962492900002804 is OK
- 10.1111/cgf.13472 is OK
- 10.7717/peerj-cs.55 is OK
- doi.org/10.18637/jss.v076.i01 is OK
- 10.1080/03640210802414826 is OK
- 10.17226/13395 is OK
- 10.1002/9781118014967 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1016/0041-55536790144-9 is INVALID
- 10.1016/s0378-47540000270-6 is INVALID
- https://doi.org/10.1016/j.jocs.2020.101204 is INVALID because of 'https://doi.org/' prefix
```

scottshambaugh committed:
Remove http from dois

scottshambaugh commented:
@mepa it looks like scoobies is stripping out the parentheses from 
* 10.1016/0041-5553(67)90144-9
* 10.1016/s0378-4754(00)00270-6

But those DOIs are good:
* https://doi.org/10.1016/0041-5553(67)90144-9
* https://doi.org/10.1016/S0378-4754(00)00270-6

This looks like an upstream bug, I'll open a report there.

mepa commented:
Thank you, @scottshambaugh!

mepa commented:
Hi @shubhamsharma1609! Do you have a final recommendation for this paper? Is it now ready for inclusion in the proceedings?

shubham-scisar commented:
@mepa Apologies for the delay from my side, I just got caught up with some work. After careful review of the recent draft ,keeping in view the comments that I had previously made regarding utilization of Monte carlo simulations in various applications  and considering that the author has pointed out the repository to carefully demonstrate the worked out examples, I think that the paper in it's current form clearly achieves it's objective. I find that the recent draft is very comprehensive and brings more clarity with the changes as suggested by @csadorf . Conclusively, I think the recent draft seems to be fit to be included in the proceedings.

mepa commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,400 @@
+
+However, basic Monte Carlo sampling is subject to the classical drawbacks of the method such as poor sampling of rare events and the slow :math:`\sigma / \sqrt{n}` convergence on quantities of interest. If the outputs and regions of interest are firmly known at the outset, then other sampling methods will be more efficient :cite:`kroese2013handbook`.
+
+Additionally, given that any conclusions are conditional on the correctness of the underlying model and input parameters, the task of validation is critical to confidence in the UA and SA results. However, this is currently out of scope for the library and must be performed with other tools. In a data-poor domain, hypothesis testing or probablistic prediction measures like loss scores can be used ...
```
typo: probablistic > probabilistic

mepa commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,400 @@
+Resource Usage
+--------------
+
+Note that *monaco*'s computational and storage overhead in creating easily-iterrogatable objects for each variable, value, and case makes it an inefficient choice for computationally simple applications with high :math:`n`, such as Monte Carlo integration. If the preprocessed sim input and raw output for each case (which for some models may dominate storage) is not retained, then the storage bott ...
```
Is "iterrogatable" is a word? Maybe a typo?

mepa commented on papers/scott_shambaugh/00_shambaugh.rst:
```diff
@@ -0,0 +1,400 @@
+Order Statistics, or, How Many Cases to Run?
+--------------------------------------------
+
+How many Monte Carlo cases should one run? One answer would be to choose :math:`n \geq 2^k` with a sampling method that implements a (t,m,s) digital net (such as a Sobol or Halton sequence), which guarentees that there will be at least one sample point in every hyperoctant of the input space :cite:`joe2008constructing`. This should be considered a lower bound for SA, with the number of cases run  ...
```
typo: guarentees -> guarantees

mepa reviewed (commented):
Hi @scottshambaugh, nice work on the paper! I marked some minor typos I noticed while looking it over. 

Thanks, @csadorf and @shubhamsharma1609 for your reviews!

mepa commented:
@scoobies mark ready

scottshambaugh committed:
Typos

scottshambaugh committed:
Final spelling and grammar pass

scottshambaugh commented:
Fixed those typos @mepa, and did a last pass for spelling/grammar. Thank you for looking it over!

scottshambaugh commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper scottshambaugh-shambaugh scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}
